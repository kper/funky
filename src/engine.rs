use crate::engine::StackContent::*;
use crate::engine::Value::*;
use std::cell::RefCell;
use std::fmt;
use std::ops::{Add, Mul};
use std::rc::{Rc, Weak};
use wasm_parser::core::CtrlInstructions::*;
use wasm_parser::core::Instruction::*;
use wasm_parser::core::NumericInstructions::*;
use wasm_parser::core::ParamInstructions::*;
use wasm_parser::core::VarInstructions::*;
use wasm_parser::core::*;
use wasm_parser::Module;

#[derive(Debug)]
pub struct Engine {
    pub module: Rc<RefCell<ModuleInstance>>, //TODO rename to `module_instance`
    pub started: bool,
    pub store: Store,
}

#[derive(Debug, PartialEq, Clone, Copy)]
#[allow(dead_code)]
pub enum Value {
    I32(i32),
    I64(i64),
    F32(f32),
    F64(f64),
}

type Arity = u32;

#[derive(Debug, Clone)]
#[allow(dead_code)]
pub enum ExtendedInstruction {
    Trap,
    Invoke(FuncIdx),
    InitElement(TableIdx, u32, Vec<FuncIdx>),
    InitData(MemoryIdx, u32, Vec<u8>),
    Label(Arity, Vec<ExtendedInstruction>, Vec<Instruction>),
    Frame(Arity, Option<Frame>, Vec<Instruction>),
}

impl Into<ValueType> for Value {
    fn into(self) -> ValueType {
        match self {
            Value::I32(_) => ValueType::I32,
            Value::I64(_) => ValueType::I64,
            Value::F32(_) => ValueType::F32,
            Value::F64(_) => ValueType::F64,
        }
    }
}

//#[derive(Debug, PartialEq, Clone, Copy)]
//pub struct Trap;
//type Result<T> = std::result::Result<T, Trap>;

impl Add for Value {
    type Output = Self;
    fn add(self, other: Self) -> Self {
        match (self, other) {
            (I32(v1), I32(v2)) => I32(v1 + v2),
            (I64(v1), I64(v2)) => I64(v1 + v2),
            (F32(v1), F32(v2)) => F32(v1 + v2),
            (F64(v1), F64(v2)) => F64(v1 + v2),
            _ => panic!("Type missmatch during addition"),
        }
    }
}
impl Mul for Value {
    type Output = Self;
    fn mul(self, other: Self) -> Self {
        match (self, other) {
            (I32(v1), I32(v2)) => I32(v1 * v2),
            (I64(v1), I64(v2)) => I64(v1 * v2),
            (F32(v1), F32(v2)) => F32(v1 * v2),
            (F64(v1), F64(v2)) => F64(v1 * v2),
            _ => panic!("Type missmatch during addition"),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Variable {
    pub mutable: bool, //Actually, there is a `Mut` enum. TODO check if makes sense to use it
    pub val: Value,
}

#[derive(Debug, Clone, PartialEq)]
#[allow(dead_code)]
pub enum StackContent {
    Value(Value),
    Frame(Frame), //TODO add labels
}

#[derive(Debug, Clone)]
pub struct Frame {
    pub arity: u32,
    pub locals: Vec<Value>,
    pub module_instance: Weak<RefCell<ModuleInstance>>,
}

impl PartialEq for Frame {
    fn eq(&self, other: &Self) -> bool {
        self.arity == other.arity && self.locals == other.locals
        //&& self.module_instance.ptr_eq(&other.module_instance)
    }
}

#[derive(Debug, Clone)]
pub struct ModuleInstance {
    pub start: u32,
    pub code: Vec<FunctionBody>,
    pub fn_types: Vec<FunctionSignature>,
    pub funcaddrs: Vec<FuncIdx>,
    pub tableaddrs: Vec<TableIdx>,
    pub memaddrs: Vec<MemoryIdx>,
    pub globaladdrs: Vec<GlobalIdx>,
    pub exports: Vec<ExportInstance>,
}

#[derive(Debug, Clone)]
pub struct Store {
    pub funcs: Vec<FuncInstance>,
    pub tables: Vec<TableInstance>,
    pub memory: Vec<MemoryInstance>,
    pub stack: Vec<StackContent>,
    pub globals: Vec<Variable>, //=GlobalInstance
}

#[derive(Debug, Clone)]
pub struct FuncInstance {
    //FIXME Add HostFunc
    pub ty: FunctionSignature,
    pub module: Weak<RefCell<ModuleInstance>>,
    pub code: FunctionBody,
}

#[derive(Debug, Clone)]
pub struct TableInstance {
    pub elem: Vec<Option<FuncIdx>>,
    pub max: Option<u32>,
}

#[derive(Clone)]
pub struct MemoryInstance {
    pub data: Vec<u8>,
    pub max: Option<u32>,
}

/// Overwritten debug implementation
/// Because `data` can have a lot of entries, which
/// can be a problem when printing
impl fmt::Debug for MemoryInstance {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("MemoryInstance")
            .field("data (only length)", &self.data.len())
            .field("max", &self.max)
            .finish()
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ExportInstance {
    pub name: String,
    pub value: ExternalKindType, //TODO maybe drop the Type in name?
}

impl Into<ExportInstance> for &ExportEntry {
    fn into(self) -> ExportInstance {
        ExportInstance {
            name: self.name.clone(),
            value: self.kind,
        }
    }
}

/*
#[derive(Debug, Clone)]
#[allow(dead_code)]
pub enum ExternalVal {
    Func(FuncIdx),
    Table(TableIdx),
    Mem(MemoryIdx),
    Global(GlobalIdx),
}
*/

macro_rules! fetch_binop {
    ($stack: expr) => {{
        let v1 = match $stack.pop().unwrap() {
            Value(v) => v,
            x => panic!("Top of stack was not of type $v_ty: {:?}", x),
        };
        let v2 = match $stack.pop().unwrap() {
            Value(v) => v,
            x => panic!("Top of stack was not of type $v_ty: {:?}", x),
        };
        (v1, v2)
    }};
}

impl ModuleInstance {
    pub fn new(m: &Module) -> Self {
        let mut mi = ModuleInstance {
            start: 0,
            code: Vec::new(),
            fn_types: Vec::new(),
            funcaddrs: Vec::new(),
            tableaddrs: Vec::new(),
            memaddrs: Vec::new(),
            globaladdrs: Vec::new(),
            exports: Vec::new(),
        };
        for section in m.sections.iter() {
            match section {
                Section::Code(CodeSection { entries: x }) => mi.code = x.clone(),
                Section::Type(TypeSection { entries: x }) => mi.fn_types = x.clone(),
                _ => {}
            }
        }

        mi
    }
}

impl Engine {
    pub fn new(mi: ModuleInstance, module: &Module) -> Self {
        let mut e = Engine {
            module: Rc::new(RefCell::new(mi)),
            started: false,
            store: Store {
                funcs: Vec::new(),
                tables: Vec::new(),
                stack: Vec::new(),
                globals: Vec::new(),
                memory: Vec::new(),
            },
        };

        e.allocate(module);

        e
    }

    pub fn downgrade_mod_instance(&self) -> Weak<RefCell<ModuleInstance>> {
        Rc::downgrade(&self.module)
    }

    fn allocate(&mut self, m: &Module) {
        info!("Allocation");
        crate::allocation::allocate(m, &self.module, &mut self.store).expect("Allocation failed");
    }

    pub fn instantiation(&mut self, m: &Module) {
        info!("Instantiation");
        crate::instantiation::instantiation(m, &self.module, &mut self.store)
            .expect("Instantiation failed");
    }

    #[warn(dead_code)]
    pub fn invoke_function(&mut self, idx: u32, args: Vec<Value>) {
        self.store.stack.push(Frame(Frame {
            arity: args.len() as u32,
            locals: args,
            module_instance: Rc::downgrade(&self.module),
        }));
        debug!("Invoking function on {:#?}", self);
        self.run_function(idx);
    }
    fn run_function(&mut self, idx: u32) {
        debug!("Running function {:?}", idx);
        //let instance = self.module.borrow_mut();
        let f = self.module.borrow().code[idx as usize].clone();
        let mut fr = match self.store.stack.last().cloned() {
            Some(Frame(fr)) => fr,
            Some(x) => panic!("Expected frame but found {:?}", x),
            None => panic!("Empty stack on function call"),
        };
        let mut ip = 0;
        while ip < f.code.len() {
            debug!("Evaluating instruction {:?}", &f.code[ip]);
            match &f.code[ip] {
                Var(OP_LOCAL_GET(idx)) => self.store.stack.push(Value(fr.locals[*idx as usize])),
                Var(OP_LOCAL_SET(idx)) => match self.store.stack.pop() {
                    Some(Value(v)) => fr.locals[*idx as usize] = v,
                    Some(x) => panic!("Expected value but found {:?}", x),
                    None => panic!("Empty stack during local.set"),
                },
                Var(OP_LOCAL_TEE(idx)) => match self.store.stack.last() {
                    Some(Value(v)) => fr.locals[*idx as usize] = *v,
                    Some(x) => panic!("Expected value but found {:?}", x),
                    None => panic!("Empty stack during local.set"),
                },
                Var(OP_GLOBAL_GET(idx)) => self
                    .store
                    .stack
                    .push(Value(self.store.globals[*idx as usize].val)),
                Var(OP_GLOBAL_SET(idx)) => match self.store.stack.pop() {
                    Some(Value(v)) => {
                        if !self.store.globals[*idx as usize].mutable {
                            panic!("Attempting to modify a immutable global")
                        }
                        self.store.globals[*idx as usize].val = v
                    }
                    Some(x) => panic!("Expected value but found {:?}", x),
                    None => panic!("Empty stack during local.set"),
                },
                Num(OP_I32_CONST(v)) => self.store.stack.push(Value(I32(*v))),
                Num(OP_I64_CONST(v)) => self.store.stack.push(Value(I64(*v))),
                Num(OP_F32_CONST(v)) => self.store.stack.push(Value(F32(*v))),
                Num(OP_F64_CONST(v)) => self.store.stack.push(Value(F64(*v))),
                Num(OP_I32_ADD) | Num(OP_I64_ADD) | Num(OP_F32_ADD) | Num(OP_F64_ADD) => {
                    let (v1, v2) = fetch_binop!(self.store.stack);
                    self.store.stack.push(Value(v1 + v2))
                }
                Num(OP_I32_MUL) | Num(OP_I64_MUL) | Num(OP_F32_MUL) | Num(OP_F64_MUL) => {
                    let (v1, v2) = fetch_binop!(self.store.stack);
                    self.store.stack.push(Value(v1 * v2))
                }
                Param(OP_DROP) => {
                    self.store.stack.pop();
                }
                Param(OP_SELECT) => {
                    let c = match self.store.stack.pop() {
                        Some(Value(I32(x))) => x,
                        _ => panic!("Expected I32 on top of stack"),
                    };
                    let (v1, v2) = fetch_binop!(self.store.stack);
                    if c != 0 {
                        self.store.stack.push(Value(v1))
                    } else {
                        self.store.stack.push(Value(v2))
                    }
                }
                Ctrl(OP_CALL(idx)) => {
                    let t = self.module.borrow().fn_types[*idx as usize].clone();
                    let args = self
                        .store
                        .stack
                        .split_off(self.store.stack.len() - t.param_types.len())
                        .into_iter()
                        .map(|x| match x {
                            Value(v) => v,
                            other => panic!("Expected value but found {:?}", other),
                        })
                        .collect();
                    let cfr = Frame {
                        arity: t.return_types.len() as u32,
                        locals: args,
                        module_instance: Rc::downgrade(&self.module),
                    };
                    debug!("Calling {:?} with {:#?}", *idx, cfr);
                    self.store.stack.push(Frame(cfr));
                    self.run_function(*idx);
                }
                Ctrl(OP_RETURN) | Ctrl(OP_END) => {
                    break;
                }
                Ctrl(OP_NOP) => {}
                x => panic!("Instruction {:?} not implemented", x),
            }
            ip += 1;
        }
        // implicit return
        let mut ret = Vec::new();
        for _ in 0..fr.arity {
            match self.store.stack.pop() {
                Some(Value(v)) => ret.push(Value(v)),
                Some(x) => panic!("Expected value but found {:?}", x),
                None => panic!("Unexpected empty stack!"),
            }
        }
        while let Some(Frame(_)) = self.store.stack.pop() {}
        self.store.stack.append(&mut ret);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn empty_engine() -> Engine {
        let mi = Rc::new(RefCell::new(ModuleInstance {
            start: 0,
            code: Vec::new(),
            fn_types: Vec::new(),
            funcaddrs: Vec::new(),
            tableaddrs: Vec::new(),
            memaddrs: Vec::new(),
            globaladdrs: Vec::new(),
            exports: Vec::new(),
        }));
        Engine {
            started: true,
            store: Store {
                funcs: Vec::new(),
                tables: Vec::new(),
                globals: Vec::new(),
                memory: Vec::new(),
                stack: vec![Frame(Frame {
                    arity: 0,
                    locals: Vec::new(),
                    module_instance: Rc::downgrade(&mi),
                })],
            },
            module: mi,
        }
    }

    #[test]
    fn test_run_function() {
        let mut e = empty_engine();
        e.store.stack = vec![Frame(Frame {
            arity: 1,
            locals: Vec::new(),
            module_instance: e.downgrade_mod_instance(),
        })];
        e.module.borrow_mut().code = vec![FunctionBody {
            locals: vec![],
            code: vec![
                Num(OP_I32_CONST(42)),
                Num(OP_I32_CONST(42)),
                Num(OP_I32_ADD),
            ],
        }];
        e.run_function(0);
        assert_eq!(Value(I32(84)), e.store.stack.pop().unwrap());
        e.store.stack = vec![Frame(Frame {
            arity: 1,
            locals: Vec::new(),
            module_instance: e.downgrade_mod_instance(),
        })];
        e.module.borrow_mut().code = vec![FunctionBody {
            locals: vec![],
            code: vec![
                Num(OP_I64_CONST(32)),
                Num(OP_I64_CONST(32)),
                Num(OP_I64_ADD),
                Num(OP_I64_CONST(2)),
                Num(OP_I64_MUL),
            ],
        }];
        e.run_function(0);
        assert_eq!(Value(I64(128)), e.store.stack.pop().unwrap());
    }

    #[test]
    fn test_function_with_params() {
        let mut e = empty_engine();
        e.store.stack = vec![Frame(Frame {
            arity: 1,
            locals: vec![I32(1), I32(4)],
            module_instance: e.downgrade_mod_instance(),
        })];
        e.module.borrow_mut().code = vec![FunctionBody {
            locals: vec![],
            code: vec![Var(OP_LOCAL_GET(0)), Var(OP_LOCAL_GET(1)), Num(OP_I32_ADD)],
        }];
        e.run_function(0);
        assert_eq!(Value(I32(5)), e.store.stack.pop().unwrap());
    }

    #[test]
    fn test_function_local_set() {
        let mut e = empty_engine();
        e.store.stack = vec![Frame(Frame {
            arity: 1,
            locals: vec![I32(1), I32(4)],
            module_instance: e.downgrade_mod_instance(),
        })];
        e.module.borrow_mut().code = vec![FunctionBody {
            locals: vec![],
            code: vec![
                Var(OP_LOCAL_GET(0)),
                Var(OP_LOCAL_GET(1)),
                Num(OP_I32_ADD),
                Var(OP_LOCAL_SET(0)),
                Num(OP_I32_CONST(32)),
                Var(OP_LOCAL_GET(0)),
                Num(OP_I32_ADD),
            ],
        }];
        e.run_function(0);
        assert_eq!(Value(I32(37)), e.store.stack.pop().unwrap());
    }

    #[test]
    fn test_function_globals() {
        let mut e = empty_engine();
        e.store.globals = vec![Variable {
            mutable: true,
            val: I32(69),
        }];
        e.module.borrow_mut().code = vec![FunctionBody {
            locals: vec![],
            code: vec![
                Var(OP_GLOBAL_GET(0)),
                Num(OP_I32_CONST(351)),
                Num(OP_I32_ADD),
                Var(OP_GLOBAL_SET(0)),
            ],
        }];
        e.run_function(0);
        assert_eq!(I32(420), e.store.globals[0].val);
    }

    #[test]
    fn test_drop_select() {
        let mut e = empty_engine();
        e.store.globals = vec![Variable {
            mutable: true,
            val: I32(20),
        }];
        e.module.borrow_mut().code = vec![FunctionBody {
            locals: vec![],
            code: vec![
                Num(OP_I32_CONST(1)),
                Num(OP_I32_CONST(2)),
                Num(OP_I32_CONST(0)),
                Num(OP_I32_CONST(4)),
                Param(OP_DROP),
                Param(OP_SELECT),
                Var(OP_GLOBAL_SET(0)),
            ],
        }];
        e.run_function(0);
        assert_eq!(I32(1), e.store.globals[0].val);
    }
}
