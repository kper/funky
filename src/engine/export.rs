use wasm_parser::core::{ExternalKindType, ExportEntry};

#[derive(Debug, Clone, PartialEq)]
pub struct ExportInstance {
    pub name: String,
    pub value: ExternalKindType,
}

impl From<&ExportEntry> for ExportInstance {
    fn from(state: &ExportEntry) -> ExportInstance {
        ExportInstance {
            name: state.name.clone(),
            value: state.kind,
        }
    }
}
