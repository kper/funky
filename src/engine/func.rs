use wasm_parser::core::{FunctionBody, FunctionSignature};

#[derive(Debug, Clone)]
pub struct FuncInstance {
    //FIXME Add HostFunc
    pub ty: FunctionSignature,
    pub code: FunctionBody,
}
