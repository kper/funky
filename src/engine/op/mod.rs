mod local_get;
mod local_set;
mod local_tee;
mod global_get;
mod global_set;
mod select;
mod memory_size;
mod memory_grow;
mod block;
mod br;
mod call;
mod call_indirect;
