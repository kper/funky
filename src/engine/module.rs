use crate::engine::export::ExportInstance;
use anyhow::{Result};
use wasm_parser::core::*;
use wasm_parser::Module;

/// The module instance contains instance information for a module.
#[derive(Debug, Default, Clone)]
pub struct ModuleInstance {
    /// Contains all functions of a module
    code: Vec<FunctionBody>,
    /// Declares the function signatures of the functions
    fn_types: Vec<FunctionSignature>,
    /// Keeps the indexes of the function
    func_addrs: Vec<FuncAddr>,
    /// Keeps the indexes of the table
    table_addrs: Vec<TableAddr>,
    /// Keeps the indexes of the memories (currently the spec only allows one memory section)
    mem_addrs: Vec<MemoryAddr>,
    /// Keeps the indexes of the global variables
    global_addrs: Vec<GlobalAddr>,
    exports: Vec<ExportInstance>,
}

impl ModuleInstance {
    pub fn new(m: &Module) -> Self {
        let mut mi = ModuleInstance {
            code: Vec::new(),
            fn_types: Vec::new(),
            func_addrs: Vec::new(),
            table_addrs: Vec::new(),
            mem_addrs: Vec::new(),
            global_addrs: Vec::new(),
            exports: Vec::new(),
        };
        for section in m.sections.iter() {
            match section {
                Section::Code(CodeSection { entries: x }) => {
                    mi.code = x.clone();
                }
                Section::Type(TypeSection { entries: x }) => {
                    mi.fn_types = x.clone();
                }
                _ => {}
            }
        }

        mi
    }

    /// Adding a new function type.
    /// We need this function to test blocks, with multiple
    /// return values.
    pub(crate) fn add_func_type(&mut self, r: Vec<ValueType>) -> usize {
        let instance = FunctionSignature {
            param_types: vec![],
            return_types: r,
        };

        self.fn_types.push(instance);

        self.fn_types.len() - 1
    }

    /// Looking up the function's address in the store by given function's module address.
    pub fn lookup_function_addr(&self, idx: &FuncIdx) -> Option<&FuncAddr> {
        self
            .func_addrs
            .get((*idx) as usize)
    }

    pub fn lookup_table_addr(&self, idx: &TableIdx) -> Option<&TableAddr> {
        self
            .table_addrs
            .get(*idx as usize)
    }

    pub fn lookup_memory_addr(&self, idx: &MemoryIdx) -> Option<&MemoryAddr> {
        self
            .mem_addrs
            .get(*idx as usize)
    }

    pub fn lookup_global_addr(&self, idx: &GlobalIdx) -> Option<&GlobalAddr> {
        self
            .global_addrs
            .get(*idx as usize)
    }

    /// Looking up the the func type with given index.
    pub fn lookup_func_types(&self, index: &u32) -> Option<&FunctionSignature> {
        self.fn_types.get(*index as usize)
    }

    /// Looking up the code with given index.
    pub fn lookup_code(&self, index: usize) -> Option<&FunctionBody> {
        self.code.get(index)
    }

    /// Storing a new function addr.
    pub fn store_func_addr(&mut self, new_addr: FuncAddr) -> Result<()> {
        self.func_addrs.push(new_addr);

        Ok(())
    }

    /// Storing a new table addr.
    pub fn store_table_addr(&mut self, new_addr: TableAddr) -> Result<()> {
        self.table_addrs.push(new_addr);

        Ok(())
    }

    /// Storing a new memory addr.
    pub fn store_memory_addr(&mut self, new_addr: MemoryAddr) -> Result<()> {
        self.mem_addrs.push(new_addr);

        Ok(())
    }

    /// Storing a new global addr.
    pub fn store_global_addr(&mut self, new_addr: GlobalAddr) -> Result<()> {
        self.global_addrs.push(new_addr);

        Ok(())
    }

    /// Storing a new export instance.
    pub fn store_export(&mut self, export: ExportInstance) -> Result<()> {
        self.exports.push(export);

        Ok(())
    }

    /// Get an immutable borrow for all memory addresses.
    pub fn get_mem_addrs(&self) -> &Vec<MemoryAddr> {
        &self.mem_addrs
    }

    /// Get the index of the export instance by iterating through all exports and filtering for the given name.
    pub fn position_export_instance_by_name(&self, name: impl Into<String>) -> Option<usize> {
        let k = name.into();
        self.exports.iter().position(|x| x.name == k)
    }

    /// Get an export instance by iterating through all exports and filtering for the given name.
    pub fn get_export_instance_by_name(&self, name: impl Into<String>) -> Option<&ExportInstance> {
        let k = name.into();
        self.exports.iter().find(|x| x.name == k)
    }

    /// Get the export instance by index.
    pub fn get_export_instance(&self, idx: usize) -> Option<&ExportInstance> {
        self.exports.get(idx)
    }

    /// Add a code.
    pub fn add_code(&mut self, body: FunctionBody) -> Result<()> {
        self.code.push(body);

        Ok(())
    }

    /// Get an immutable borrow of the code.
    pub fn get_code(&self) -> &[FunctionBody] {
        &self.code
    }

    pub fn get_fn_types(&self) -> &[FunctionSignature] {
        &self.fn_types
    }

    pub fn get_func_addrs(&self) -> &[FuncAddr] {
        &self.func_addrs
    }

    pub fn get_table_addrs(&self) -> &[TableAddr] {
        &self.table_addrs
    }

    pub fn get_global_addrs(&self) -> &[GlobalAddr] {
        &self.global_addrs
    }

    pub fn get_exports(&self) -> &[ExportInstance] {
        &self.exports
    }
}
