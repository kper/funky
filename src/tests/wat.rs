use crate::engine::module::ModuleInstance;
use crate::engine::*;
use crate::value::Value::*;
use insta::assert_snapshot;
use std::fs::File;
use std::io::{self, Write};
use validation::validate;
use wasm_parser::{parse, read_wasm};

macro_rules! wat {
    ($name:expr, $input:expr, $invoke:expr, $init:expr) => {{
        use std::process::Command;
        use tempfile::tempdir;

        let tmp_dir = tempdir().unwrap();
        let file_path = tmp_dir.path().join("file.wat");

        {
            let mut tmp_file = File::create(file_path.clone()).unwrap();
            writeln!(tmp_file, $input).unwrap();
            tmp_file.flush().unwrap();

            println!("tmp_file {}", $input);
        }

        let output = Command::new("wat2wasm")
            .arg(file_path)
            .arg("-o")
            .arg(tmp_dir.path().join("file.wasm").to_str().unwrap())
            .output()
            .expect("Command failed");

        io::stdout().write_all(&output.stdout).unwrap();
        io::stderr().write_all(&output.stderr).unwrap();

        assert!(output.status.success());

        let file_path = tmp_dir.path().join("file.wasm");

        // Read it
        let file = read_wasm!(file_path);
        let module = parse(file).expect("Parsing failed");
        assert!(validate(&module).is_ok());

        let imports = Vec::new();

        let (instance, functions) = ModuleInstance::new(&module);
        let mut engine = Engine::new(
            instance,
            &functions,
            &module,
            Box::new(crate::debugger::RelativeProgramCounter::default()),
            &imports,
        )
        .unwrap();

        if let Err(err) = engine.invoke_exported_function_by_name($invoke, $init) {
            error!("ERROR: {}", err);
            err.chain()
                .skip(1)
                .for_each(|cause| error!("because: {}", cause));

            panic!("Test failed");
        }

        assert_snapshot!($name, format!("{:#?}", engine));

        engine
    }};
}

#[test]
fn test_block() {
    wat!(
        "test_block",
        "(module (func (result i32) (block (result i32) i32.const 1)) (export \"main\" (func 0)))",
        "main",
        vec![]
    );
}

#[test]
fn test_block_add() {
    wat!("test_block_add", "(module (func (result i32) (block (result i32) i32.const 1 i32.const 1 i32.add)) (export \"main\" (func 0)))", "main", vec![]);
}

#[test]
fn test_call_indirect_mid() {
    wat!(
        "indirect_mid",
        "
    (module  
    (func $func (param i32 i32) (result i32) (local.get 0))
    (type $check (func (param i32 i32) (result i32)))
    (table funcref (elem $func))
    (func (export \"as-call_indirect-mid\") (param i32) (result i32)
    (block (result i32)
      (call_indirect (type $check)
        (i32.const 1) (select (i32.const 2) (i32.const 3) (local.get 0)) (i32.const 0)
      )
    ))
  )",
        "as-call_indirect-mid",
        vec![I32(1)]
    );
}

#[test]
fn test_nested_br_value() {
    wat!(
        "nested_br_value",
        "(module (func (export \"nested-br_table-value-index\") (result i32)
    (i32.add
      (i32.const 1)
      (block (result i32)
        (drop (i32.const 2))
        (br_table 0 (i32.const 4) (br 0 (i32.const 8)))
        (i32.const 16)
      )
    )
  ))",
        "nested-br_table-value-index",
        vec![]
    );
}

#[test]
fn test_br_table_index() {
    wat!(
        "br_table_index",
        "(module 
    (func (export \"nested-br_table-value-index\") (result i32)
    (i32.add
      (i32.const 1)
      (block (result i32)
        (drop (i32.const 2))
        (br_table 0 (i32.const 4) (br 0 (i32.const 8)))
        (i32.const 16)
      )))   
    )",
        "nested-br_table-value-index",
        vec![]
    );
}

#[test]
fn test_br_table_value() {
    wat!(
        "br_table_value",
        "(module 
     (func (export \"nested-br_table-value\") (result i32)
    (i32.add
      (i32.const 1)
      (block (result i32)
        (drop (i32.const 2))
        (drop
          (block (result i32)
            (drop (i32.const 4))
            (br_table 0 (br 1 (i32.const 8)) (i32.const 1))
          )
        )
        (i32.const 16)
      )
    )) 
    )",
        "nested-br_table-value",
        vec![]
    );
}

#[test]
fn test_nested_block_value() {
    wat!(
        "nested_block_value",
        "(module
     (func (export \"nested-block-value\") (param i32) (result i32)
    (block (result i32)
      (drop (i32.const -1))
      (i32.add
        (i32.const 1)
        (block (result i32)
          (i32.add
            (i32.const 2)
            (block (result i32)
              (drop (i32.const 4))
              (i32.add
                (i32.const 8)
                (br_table 0 1 2 (i32.const 16) (local.get 0))
              )
            )
          )
        )))))",
        "nested-block-value",
        vec![I32(1)]
    );
}

#[test]
fn test_as_if_then() {
    wat!(
        "as-if-then",
        "(module
    (func (export \"as-if-then\") (param i32 i32) (result i32)
    (block (result i32)
      (if (result i32)
        (local.get 0)
        (then (br_table 1 (i32.const 3) (i32.const 0)))
        (else (local.get 1))
      )
    )
  ))",
        "as-if-then",
        vec![I32(1), I32(6)]
    );
}

#[test]
fn test_br_table_num_num() {
    wat!(
        "break-br_table_num_num",
        "(module
     (func (export \"break-br_table-num-num\") (param i32) (result i32 i64)
        (br_table 0 0 (i32.const 50) (i64.const 51) (local.get 0))
        (i32.const 51) (i64.const 52)
    ))",
        "break-br_table-num-num",
        vec![I32("4294967196".parse::<u32>().unwrap() as i32)]
    );
}

#[test]
fn test_params_break_id() {
    wat!(
        "params_break_id",
        "(module (func (export \"params-id-break\") (result i32)
    (local $x i32)
    (local.set $x (i32.const 0))
    (i32.const 1)
    (i32.const 2)
    (loop (param i32 i32) (result i32 i32)
      (local.set $x (i32.add (local.get $x) (i32.const 1)))
      (br_if 0 (i32.lt_u (local.get $x) (i32.const 10)))
    )
    (i32.add)
  ))",
        "params-id-break",
        vec![]
    );
}

#[test]
fn test_fac_ssa() {
    wat!(
        "fac_ssa",
        "(module (func $pick0 (param i64) (result i64 i64)
    (local.get 0) (local.get 0)
  )
  (func $pick1 (param i64 i64) (result i64 i64 i64)
    (local.get 0) (local.get 1) (local.get 0)
  )
  (func (export \"fac-ssa\") (param i64) (result i64)
    (i64.const 1) (local.get 0)
    (loop $l (param i64 i64) (result i64)
      (call $pick1) (call $pick1) (i64.mul)
      (call $pick1) (i64.const 1) (i64.sub)
      (call $pick0) (i64.const 0) (i64.gt_u)
      (br_if $l)
      (drop) (return)
    )
  ))",
        "fac-ssa",
        vec![I64(3)]
    );
}

#[test]
fn test_as_return_values() {
    wat!(
        "as_return_values",
        "(module
    (func (export \"as-return-values\") (result i32 i64)
    (i32.const 2)
    (block (result i64) (return (br 0 (i32.const 1) (i64.const 7))))
  ))
    ",
        "as-return-values",
        vec![]
    );
}

#[test]
fn test_add64_sat_u() {
    wat!(
        "add64_sat_u",
        "(module
   (func $add64_u_with_carry (export \"add64_u_with_carry\")
    (param $i i64) (param $j i64) (param $c i32) (result i64 i32)
    (local $k i64)
    (local.set $k
      (i64.add
        (i64.add (local.get $i) (local.get $j))
        (i64.extend_i32_u (local.get $c))
      )
    )
    (return (local.get $k) (i64.lt_u (local.get $k) (local.get $i)))
  )

  (func $add64_u_saturated (export \"add64_u_saturated\")
    (param i64 i64) (result i64)
    (call $add64_u_with_carry (local.get 0) (local.get 1) (i32.const 0))
    (if (param i64) (result i64)
      (then (drop) (i64.const -1))
    )
  ) 
) 
    ",
        "add64_u_saturated",
        vec![
            I64("9223372036854775808".parse::<u64>().unwrap() as i64),
            I64("9223372036854775808".parse::<u64>().unwrap() as i64)
        ]
    );
}
