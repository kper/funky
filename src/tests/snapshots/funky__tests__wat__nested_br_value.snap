---
source: src/tests/wat.rs
expression: "format!(\"{:#?}\", engine)"

---
Engine {
    module_instance: ModuleInstance {
        fn_types: [
            FunctionSignature {
                param_types: [],
                return_types: [
                    I32,
                ],
            },
        ],
        func_addrs: [
            FuncAddr(
                0,
            ),
        ],
        table_addrs: [],
        mem_addrs: [],
        global_addrs: [],
        exports: [
            ExportInstance {
                name: "nested-br_table-value-index",
                value: Function {
                    ty: 0,
                },
            },
        ],
    },
    started: false,
    store: Store {
        funcs: [
            FuncInstance {
                ty: FunctionSignature {
                    param_types: [],
                    return_types: [
                        I32,
                    ],
                },
                code: FunctionBody {
                    locals: [],
                    code: [
                        InstructionWrapper {
                            instruction_id: 8,
                            instruction: OP_I32_CONST(
                                1,
                            ),
                        },
                        InstructionWrapper {
                            instruction_id: 9,
                            instruction: OP_BLOCK(
                                ValueType(
                                    I32,
                                ),
                                CodeBlock {
                                    id: 1,
                                    instructions: [
                                        InstructionWrapper {
                                            instruction_id: 1,
                                            instruction: OP_I32_CONST(
                                                2,
                                            ),
                                        },
                                        InstructionWrapper {
                                            instruction_id: 2,
                                            instruction: OP_DROP,
                                        },
                                        InstructionWrapper {
                                            instruction_id: 3,
                                            instruction: OP_I32_CONST(
                                                4,
                                            ),
                                        },
                                        InstructionWrapper {
                                            instruction_id: 4,
                                            instruction: OP_I32_CONST(
                                                8,
                                            ),
                                        },
                                        InstructionWrapper {
                                            instruction_id: 5,
                                            instruction: OP_BR(
                                                0,
                                            ),
                                        },
                                        InstructionWrapper {
                                            instruction_id: 6,
                                            instruction: OP_BR_TABLE(
                                                [],
                                                0,
                                            ),
                                        },
                                        InstructionWrapper {
                                            instruction_id: 7,
                                            instruction: OP_I32_CONST(
                                                16,
                                            ),
                                        },
                                    ],
                                },
                            ),
                        },
                        InstructionWrapper {
                            instruction_id: 10,
                            instruction: OP_I32_ADD,
                        },
                    ],
                },
            },
        ],
        tables: [],
        memory: [],
        stack: [
            Value(
                I32(
                    9,
                ),
            ),
        ],
        globals: [],
    },
    debugger: RelativeProgramCounter(
        10,
    ),
}
